src/middlewared/middlewared/plugins/zfs.py

```diff
@@ -952,47 +952,21 @@ def query(self, filters=None, options=None):
                 filter_getattrs(filters).issubset({'name', 'pool'})
             )
         ):
-            # Using zfs list -o name is dozens of times faster than py-libzfs
-            cmd = ['zfs', 'list', '-H', '-o', 'name', '-t', 'snapshot']
-            order_by = options.get('order_by')
-            # -s name makes it even faster
-            if not order_by or order_by == ['name']:
-                cmd += ['-s', 'name']
-            cp = subprocess.run(
-                cmd,
-                stdout=subprocess.PIPE,
-                stderr=subprocess.PIPE,
-                universal_newlines=True,
-            )
-            if cp.returncode != 0:
-                raise CallError(f'Failed to retrieve snapshots: {cp.stderr}')
-            stdout = cp.stdout.strip()
-            if not stdout:
-                return []
-            snaps = [
-                {'name': i, 'pool': i.split('/', 1)[0]}
-                for i in stdout.split('\n')
-            ]
-            if filters:
+            with libzfs.ZFS() as zfs:
+                snaps = zfs.snapshots_serialized(['name'])
+
+            if filters or len(options) > 1:
                 return filter_list(snaps, filters, options)
             return snaps
+
         with libzfs.ZFS() as zfs:
             # Handle `id` filter to avoid getting all snapshots first
-            snapshots = []
+            kwargs = dict(holds=False, mounted=False)
             if filters and len(filters) == 1 and list(filters[0][:2]) == ['id', '=']:
-                try:
-                    snapshots.append(zfs.get_snapshot(filters[0][2]).__getstate__())
-                except libzfs.ZFSException as e:
-                    if e.code != libzfs.Error.NOENT:
-                        raise
-            else:
-                for i in zfs.snapshots:
-                    try:
-                        snapshots.append(i.__getstate__())
-                    except libzfs.ZFSException as e:
-                        # snapshot may have been deleted while this is running
-                        if e.code != libzfs.Error.NOENT:
-                            raise
+                kwargs['datasets'] = [filters[0][2]]
```

Comment:
Why datasets? We want a specific snapshot id, isnt the old code more suitable for this?